examples/pagerank.cpp

```diff
@@ -0,0 +1,181 @@
+
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "boost/tokenizer.hpp"
+
+#include "core/engine.hpp"
+#include "io/input/hdfs_line_inputformat.hpp"
+
+class PRVertex {
+public:
+    float pr;
+    typedef int KeyT;
+    int key;
+
+    PRVertex() {}
+    explicit PRVertex(KeyT key) {
+        this->key = key;
+    }
+
+    inline virtual KeyT const & id() const {
+        return key;
+    }
+
+    virtual void key_init(KeyT key) {
+        this->key = key;
+    }
+
+    virtual void add_neighbors(KeyT nbKey) {
+        this->nbs.push_back(nbKey);
+    }
+
+    virtual std::vector<KeyT> & get_neighbors() {
+        return this->nbs;
+    }
+
+    static uint64_t partition(const KeyT & key) {
+        return key;
+    }
+
+    friend husky::BinStream & operator >> (husky::BinStream & stream, PRVertex & v) {
+        stream >> v.pr >> v.nbs >> v.key;
+        return stream;
+    }
+
+    friend husky::BinStream & operator << (husky::BinStream & stream, PRVertex v) {
+        stream << v.pr << v.nbs << v.key;
+        return stream;
+    }
+
+protected:
+    std::vector<KeyT> nbs;
+};
+
+std::function<void(PRVertex &)> exec_pagerank(husky::ObjList<PRVertex> & obj_list, husky::PushCombinedChannel<float, PRVertex, husky::SumCombiner<float>> & ch) {
+    std::function<void(PRVertex &)> exec_lambda = [&](PRVertex & v) {
+        float sum = 0;
+        sum += ch.get(v);
+        v.pr = 0.15 + 0.85 * sum;
+
+        if (v.get_neighbors().size() == 0) return;
+
+        float pr_contrib = v.pr / v.get_neighbors().size();
+        for (auto & nb : v.get_neighbors()) {
+            ch.push(pr_contrib, nb);
+        }
+    };
+    return exec_lambda;
+}
+
+class Foo {
+public:
+    typedef int KeyT;
+    int key;
+
+    explicit Foo(KeyT & key) {
+        this->key = key;
+    }
+
+    const KeyT & id() const {
+        return key;
+    }
+};
+
+void pagerank() {
+    io::HDFSLineInputFormat infmt;
+    infmt.set_input(husky::Context::get_param("input"));
+    husky::ObjList<PRVertex> pr_list;
+    auto & ch = husky::ChannelFactory::create_push_combined_channel<float, husky::SumCombiner<float>>(pr_list, pr_list);
+    auto parse_pr = [&](boost::string_ref & chunk) {
+        if (chunk.size() == 0) return;
+        boost::char_separator<char> sep(" \t");
+        boost::tokenizer<boost::char_separator<char>> tok(chunk, sep);
+        PRVertex obj;
+        int i = 0;
+        for (auto & v : tok) {
+            i++;
+            if (i == 1)
+                obj.key_init(std::stoi(v));
+            else if (i == 2);
+            else
+                obj.add_neighbors(std::stoi(v));
+        }
+        pr_list.add_object(obj);
+    };
+
+    husky::load(infmt, parse_pr);
+    husky::globalize(pr_list);
+    husky::list_execute(pr_list, [&](PRVertex & v) {
+        // send dummy page rank values
+        for (auto & nb : v.get_neighbors()) {
+            ch.push(0, nb);
+        }
+    });
```

Comment:
If I remove this list_execute, it will trigger segmentation fault because during the first iteration of the exec_pagerank, the worker will try to get from the channel when in fact there is nothing to get from.